warehouse/query-core/src/main/java/datawave/query/jexl/visitors/DelayedNonEventSubTreeVisitor.java

```diff
@@ -0,0 +1,165 @@
+package datawave.query.jexl.visitors;
+
+import datawave.query.jexl.JexlASTHelper;
+import org.apache.commons.jexl2.parser.ASTAndNode;
+import org.apache.commons.jexl2.parser.ASTEQNode;
+import org.apache.commons.jexl2.parser.ASTFunctionNode;
+import org.apache.commons.jexl2.parser.ASTGENode;
+import org.apache.commons.jexl2.parser.ASTGTNode;
+import org.apache.commons.jexl2.parser.ASTIdentifier;
+import org.apache.commons.jexl2.parser.ASTJexlScript;
+import org.apache.commons.jexl2.parser.ASTLENode;
+import org.apache.commons.jexl2.parser.ASTLTNode;
+import org.apache.commons.jexl2.parser.ASTNENode;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.commons.jexl2.parser.JexlNodes;
+import org.apache.log4j.Logger;
+
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+/**
+ * Visitor builds a list of all delayed sub trees within a script, and removes any nodes which are not based on index only lookups. This visitor will
+ * modify/destroy the tree so a copy should always be used
+ */
+public class DelayedNonEventSubTreeVisitor extends BaseVisitor {
```

Comment:
So, you get all the delayed subtrees in this visitor, but then you are only using the leaf nodes (which can contain includes and excludes) when this makes it's way into DelayedNonEventIndexContext.  Why not just get the leaf nodes at this point.  Then you don't need to carry subtrees all the way through.  Also, you could create a multimap of field name to leaf nodes which would be useful later on when you need to do a lookup (i.e. no need to check all delayed subtrees to find what you're looking for).  You also wouldn't have to worry about pruning the tree if you took this approach.